### ISSUE TITLE
Classes decorated with ``@object.__new__`` aren't handled correctly

### ISSUE THREAD

**Issue by DanielNoord** [microsoft/pyright#3641]
**Describe the bug**

Classes decorated with ``@object.__new__`` are not seen as an instance of their type but as their type like normal class definitions are.

**To Reproduce**
```python
@object.__new__
class A:
    ...

def func() -> type[A]:
    return A

print(func())
print(type(func()))
```

Returns:
```python
<__main__.A object at 0x10de72890>
<class '__main__.A'>
```

This shows that due to the decorator `A` is an instance of `type A`. I thought that the return type of `func()` should therefore be `type[A]`, as that is how you normally type functions that return an instance of a type.

Currently this raises:
```console
test.py
test.py:5:20 - error: Expected class type but received "A" (reportGeneralTypeIssues)
test.py:6:12 - error: Expression of type "A" cannot be assigned to return type "type"
    "A" is incompatible with "type" (reportGeneralTypeIssues)
```

**Expected behavior**

No errors.

**Screenshots or Code**

/

**VS Code extension or command-line**
Both CLI and VS Code extension report this problem.

**Additional context**

**Comment by DanielNoord** [comment]
Thanks for taking the time to respond with additional examples. If you don't mind I have two follow up questions.

> Type annotations must refer to types (classes), not instances. It is illegal to use instances in a type annotation, so the return type annotation `type[A]` is invalid. It would be similar to saying `type[1]` or `type["str"]`.

But what if I import `A` from a package I don't control, which uses the decorator without the `_A` refactor. So:

```python
from mod import A

def func() -> type[A]:
    return A
```

What should the return type of `func()` be? 

> Second, the annotation `X` is the proper way to refer to an instance of class `X`. The annotation `type[X]` refers to the class `X`. Keeping that in mind, the correct annotation here would be:
> 
> ...

I understand that that would be an ideal refactor, but that would not be the same and I'm worried about potentially breaking downstream dependents. Currently `A` is an instance of the type of `A` (if that makes sense), refactoring like this would make it an instance of `_A`. Wouldn't that require them to update their typing/code?
